packages/agents/src/functions/RunTestPythonAnalyser.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  Agent,
+  AgentFunctionResult,
+  WrapClient,
+  LlmApi,
+  ChatLogs,
+  Tokenizer,
+  ChatMessageBuilder,
+} from "@evo-ninja/agent-utils";
+import { AgentFunctionBase } from "../AgentFunctionBase";
+import { AgentBaseContext } from "../AgentBase";
+
+interface FunctionParams {
+  filename: string;
+  implementationCode: string;
+}
+
+export class RunTestPythonAnalyser extends AgentFunctionBase<FunctionParams> {
+  constructor(
+    private _llm: LlmApi,
+    private _tokenizer: Tokenizer,
+    private client: WrapClient
+  ) {
+    super();
+  }
+
+  get name() {
+    return "runPythonTest";
+  }
+
+  get description() {
+    return `Run python test and analyses error.`;
+  }
+
+  get parameters() {
+    return {
+      type: "object",
+      properties: {
+        filename: {
+          type: "string",
+        },
+        implementationCode: {
+          type: "string"
+        }
+      },
+      required: ["filename", "implementationCode"],
+      additionalProperties: false,
+    };
+  }
+
+  buildExecutor(_: Agent<unknown>, context: AgentBaseContext) {
+    return async (params: FunctionParams, rawParams?: string): Promise<AgentFunctionResult> => {
+      const testResult = await this.client.invoke<
```

Comment:
no need to use the client here, just implement the `workspace.shellExec` logic here.